### PR TITLE
Added torch.autograd.profiler.record_function() as context manager.

### LINKED ISSUE #19422: Document that autograd::Profiler::RecordFunction is available in Python
## 🚀 Feature

`torch.autograd.profiler.profile().export_chrome_trace` is great. Adding information to it from Python would make it even better.

## Motivation

In multi-threaded PyTorch code, it can be useful to further annotate the tracing output by adding additional information to it.

## Pitch

One easy / cheap way of doing this:

```cc
PYBIND11_MODULE(torch, m) {
  py::class_<torch::autograd::profiler::RecordFunction>(m, "Record")
      .def(py::init<std::string>(), py::arg("name"));
}
```

And then in Python

```python
record = torch.Record("model")
outputs = model(frame)
del record  # Alternatively, turn it into a context manager.
```

### PR DESCRIPTION
**Pull request by prasunanand**
Added torch.autograd.profiler.record_function() as context manager to annotate block of Python code during profiling.

Fixes https://github.com/pytorch/pytorch/issues/19422 .

### CODE REVIEW

**Comment by ezyang** [comment]
cool! cc @heiner

**Comment by ezyang** [comment]
This looks fine to me. @apaszke you want to take a second look? Otherwise I'll go ahead and merge it in a day.

File: torch/autograd/profiler.py [hunk 1]
```diff
@@ -308,6 +308,47 @@ def self_cpu_time_total(self):
...
+
+    Example:
+        >>> x = torch.randn((1, 1), requires_grad=True)
+        >>> with torch.autograd.profiler.profile() as prof:
+        ...     y = x ** 2
+        ...     with torch.autograd.profiler.record_function("label-z"): // label the block
+        ...         z = y ** 3
+        ...     y.backward()
+        ...
+        >>> # NOTE: some columns were removed for brevity
```
Comment (bddppq) [inline torch/autograd/profiler.py]: Side topic, shall we avoid printing those cuda columns when use_cuda is `False`?

File: torch/autograd/profiler.py [hunk 2]
```diff
@@ -308,6 +308,47 @@ def self_cpu_time_total(self):
...
+    profiler.  It is useful when tracing the code profile.
+
+    Arguments:
+        name (str): Label assigned to the block of code.
+
+    Example:
+        >>> x = torch.randn((1, 1), requires_grad=True)
+        >>> with torch.autograd.profiler.profile() as prof:
+        ...     y = x ** 2
+        ...     with torch.autograd.profiler.record_function("label-z"): // label the block
```
Comment (rgommers) [inline torch/autograd/profiler.py]: minor: a C++ style comment crept in here, should be `# label the block`

File: torch/autograd/profiler.py [hunk 3]
```diff
@@ -308,6 +308,47 @@ def self_cpu_time_total(self):
...
+        ...         z = y ** 3
+        ...     y.backward()
+        ...
+        >>> # NOTE: some columns were removed for brevity
+        ... print(prof.key_averages().table(sort_by="self_cpu_time_total"))
```
Comment (rgommers) [inline torch/autograd/profiler.py]: `...` should be `>>>`

**Review by ezyang — CHANGES_REQUESTED** [review]
There are some reviews on the example code which should get addressed

File: test/test_autograd.py [hunk 4]
```diff
@@ -2808,6 +2809,22 @@ def test_profiler_aggregation_lstm(self):
             with tempfile.NamedTemporaryFile() as trace_file:
                 prof.export_chrome_trace(trace_file.name)
 
+    def test_record_function(self):
+        x = torch.randn(10, 10)
+
+        with profile() as p:
+            with record_function("label"):
```
Comment (fmassa) [inline test/test_autograd.py]: What happens if we have multiple levels of context managers? I suppose only the name of the inner-most one appears?

I was wondering if it would make sense to also nest the names of the contexts, so that users can easily understand and parse profiles in arbitrary places of code (like `classifier.block1` instead of just `block1`)

Comment (apaszke) [reply test/test_autograd.py]: Nope, you'll see them as nested ranges in the trace.

**Review by apaszke — APPROVED** [review]
LGTM modulo the broken example

**Comment by prasunanand** [comment]
@ezyang I have added the changes.

**Review by colesbury — CHANGES_REQUESTED** [review]
I don't think the documentation will be included unless you add it to [docs/source/autograd.rst](https://github.com/pytorch/pytorch/blob/master/docs/source/autograd.rst)

**Comment by prasunanand** [comment]
Documentation gets generated now.

**Review by soumith — APPROVED** [review]
